pkg/gomod/selector.go

```diff
@@ -15,90 +14,49 @@ var (
 )
 
 // Matcher matches a module name.
-type Matcher interface {
-	// Match returns true if the module name matches.
-	Match(modname string) bool
-}
-
-// MatcherFunc is a function that implements the Matcher interface.
-func MatcherFunc(fn matcherFunc) Matcher {
-	return fn
-}
-
-type matcherFunc func(modname string) bool
-
-func (fn matcherFunc) Match(modname string) bool {
-	return fn(modname)
-}
+type Matcher func(modname string) bool
 
-// Selector is a set of matchers that can be used to select a subset of modules.
-type Selector struct {
-	Includes []Matcher
-	Excludes []Matcher
-}
-
-// Select returns true if the module should be selected.
-func (s Selector) Select(modname string) bool {
-	selected := false
-	for _, m := range s.Includes {
-		if m.Match(modname) {
-			selected = true
-			break
+// CurrentModulesMatcher matches the current project modules.
+func CurrentModulesMatcher() (Matcher, error) {
+	os := gowork.RealSystem{}
+	knownMods := sets.NewString()
+	mods, err := gowork.List(os, os)
```

Comment:
As I said in #3696 such refactor will remove the laziness, which was being done using `sync.Once`.

Now all work is done in factory function `CurrentModulesMatcher() (Matcher, error)`, which is undesired, and unexpected behavior. Factory function shouldn't do a real costly work, especially if it isn't required.

If you really like to pursuit this refactor, I would advise changing the `Matcher` func to be:

```golang
type Matcher func(modname string) (bool, error)
```

That would allow moving the "work" to the interior of the matcher func, keeping the current lazy behavior.

Reply:
Are we taking advantage of that laziness anywhere?

The advantage to doing the work up-front is that `Matcher` can have a simpler signature that doesn't need to report errors, as we do the possibly-error-creating work up front.  As far as I can tell, `CurrentModulesMatcher` is actually only ever used in this file, so we could even avoid exporting it at all.

(Yes, we could make this lazy, but I'm concerned about deferring the error to a place where it's more difficult to handle.)

Reply:
Take this with a grain of salt because of my ignorance of golang best practices and all, but for me readability > almost anything.

Reply:
Okay. I think I'm convinced by you both. This lazy behavior isn't needed. It could be easily, re-introduced if it became a problem.

Also, removal of this `log.Fatal` is a great idea.